docs/design/perfect_negotiation.md

```diff
@@ -0,0 +1,31 @@
+# Perfect Negotiation
+
+This document describes how perfect negotiation is implemented in OWT P2P SDK to avoid collision.
+
+**Perfect negotiation**, **polite peer**, **impolite peer** are defined in [Perfect negotiation example of WebRTC 1.0](https://w3c.github.io/webrtc-pc/#perfect-negotiation-example).
+
+## Determining polite peer and impolite peer
+
+OWT client SDKs determines polite peer and impolite peer by comparing client IDs. Sorting these two client IDs alphabetically in increasing order, the first one is polite order, the second one is impolite peer.
+
+Signaling server is required to return a client ID assigned for the client connected after authentication. OWT doesn't define how authentication works and how client IDs are assigned. They depend on application's design. But every client gets a unique client ID after connecting to the signaling server.
+
+## Connections
+
+We expect only one `PeerConnection` between two endpoints. A random ID is generated when creating a new `PeerConnection`. All messages (offers, answers, ICE candidates) for this `PeerConnection` have a `connectionId` property with the connection ID as its value.
+
+## Behaviors
+
+When WebRTC collision occurred, it basically follows the perfect negotiation example in WebRTC 1.0. This section only describes the implementation for signaling collision.
+
+A connection typically ends by calling `stop` at local side or receiving a `chat-closed` message from the remote side. Client SDKs stores the most recent connection IDs with all remote endpoints, and clean one of them when a connection ends. If the connection ID of a signaling message received is different from the one stored locally, collision happens.
+
+An example is both sides call `send` to create a data channel and send a message to the remote endpoint at the same time. Each side creates a new `PeerConnection`, and a connection ID. These two connection IDs are different. Then each of them will receive signaling messages with connection ID differ from the local one.
+
+### Polite peer
+
+The polite peer is the controlled side. When a signaling message with a new connection ID is received, it stops current PeerConnection， create a new one, and associate it with the remote connection ID.
+
+### Impolite peer
+
+The polite peer is the controlling side. It ignores remote messages conflict with its own state, and continue its process.
```

Comment:
conflicting with its own state, and continues its own process.

Reply:
Fixed.